connman/src/service.c

```diff
@@ -1896,18 +1879,66 @@ void __connman_service_nameserver_add_routes(struct connman_service *service,
 		 * tries to reach them. The subnet route is installed
 		 * when setting the interface IP address.
 		 */
-		nameserver_add_routes(index, service->nameservers, gw);
+		nameserver_add_routes(index, service->nameservers, gw,
+								gw_family);
 	}
 }
 
 void __connman_service_nameserver_del_routes(struct connman_service *service,
 					enum connman_ipconfig_type type)
 {
 	int index;
+	int refcount4 = -1;
+	int refcount6 = -1;
 
 	if (!service)
 		return;
 
+	DBG("service %p type %s", service,
+					__connman_ipconfig_type2string(type));
+
+	if (type != CONNMAN_IPCONFIG_TYPE_IPV6) {
+		if (service->nameservers_ipv4_refcount)
+			refcount4 = __sync_fetch_and_sub(
+					&service->nameservers_ipv4_refcount, 1);
+		else
+			refcount4 = 0;
+
+		DBG("%p IPv4 nameservers refcount %d", service, refcount4);
+	}
+
+	if (type != CONNMAN_IPCONFIG_TYPE_IPV4) {
+		if (service->nameservers_ipv6_refcount)
+			refcount6 = __sync_fetch_and_sub(
+					&service->nameservers_ipv6_refcount, 1);
+		else
+			refcount6 = 0;
+
+		DBG("%p IPv6 nameservers refcount %d", service, refcount6);
+	}
+
+	if (type == CONNMAN_IPCONFIG_TYPE_ALL &&
+					(refcount4 != -1 && refcount4 != 1) &&
+					(refcount6 != -1 && refcount6 != 1))
+		return;
+
+	if (refcount4 != -1 && refcount4 != 1) {
+		if (type == CONNMAN_IPCONFIG_TYPE_ALL)
+			type = CONNMAN_IPCONFIG_TYPE_IPV6;
+		else
+			return;
+	}
+
+	if (refcount6 != -1 && refcount6 != 1) {
+		if (type == CONNMAN_IPCONFIG_TYPE_ALL) {
+			type = CONNMAN_IPCONFIG_TYPE_IPV4;
+		} else
```

Comment:
nit: braces for 'if' but not for 'else'

Reply:
Ah, forgot to clean up this one..

Reply:
change to [5f49f52](https://github.com/sailfishos/connman/pull/55/commits/5f49f5274da18de3a7193cfb058bcaf6e276fa05) cleans this up.